File: VB/TreeList_ReorderNodes_VB/Controllers/HomeController.vb
Imports System.Web.Mvc

Namespace Controllers
    Public Class HomeController
        Inherits Controller

        Public Function Index() As ActionResult
            If Session("model") Is Nothing Then
                Dim table As New DataTable()
                table = DepartmentsProvider.GetDeparments()
                table.Columns.Add("SortIndex")

                Session("model") = table
            End If
            Return View()
        End Function

        Public Function TreeListPartial() As ActionResult
            Return PartialView(Session("model"))
        End Function

        Public Function NodeDragDropPartial(ByVal id As Integer, ByVal parentid? As Integer) As ActionResult
            Dim table As DataTable = TryCast(Session("model"), DataTable)

            Dim row1() As DataRow = table.Select(String.Format("ID = '{0}'", id.ToString()))
            Dim row2() As DataRow = table.Select(String.Format("ID = '{0}'", parentid.ToString()))
            Dim sortIndex1 As String = row1(0)("SortIndex").ToString()
            Dim sortIndex2 As String = row2(0)("SortIndex").ToString()
            row1(0)("SortIndex") = sortIndex2
            row2(0)("SortIndex") = sortIndex1

            Session("model") = table

            Return PartialView("TreeListPartial", Session("model"))
        End Function
    End Class
End Namespace

File: CS/TreeList_ReorderNodes/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using TreeList_ReorderNodes.Models;

namespace TreeList_ReorderNodes.Controllers {
    public class HomeController : Controller {

        public ActionResult Index() {
            if (Session["model"] == null) {
                DataTable table = new DataTable();
                table = DepartmentsProvider.GetDeparments();
                table.Columns.Add("SortIndex");

                Session["model"] = table;
            }
            return View();
        }

        public ActionResult TreeListPartial() {
            return PartialView(Session["model"]);
        }

        public ActionResult NodeDragDropPartial(int id, int? parentid) {
            DataTable table = Session["model"] as DataTable;

            DataRow[] row1 = table.Select(string.Format("ID = '{0}'", id.ToString()));
            DataRow[] row2 = table.Select(string.Format("ID = '{0}'", parentid.ToString()));
            string sortIndex1 = row1[0]["SortIndex"].ToString();
            string sortIndex2 = row2[0]["SortIndex"].ToString();
            row1[0]["SortIndex"] = sortIndex2;
            row2[0]["SortIndex"] = sortIndex1;

            Session["model"] = table;

            return PartialView("TreeListPartial", Session["model"]);
        }
    }
}